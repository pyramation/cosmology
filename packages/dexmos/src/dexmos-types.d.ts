export interface DenomUnits {
  denom: string
  exponent: number
  aliases?: string[]
}

export interface Coin {
  description?: string
  denom_units: DenomUnits[]
  base: string // this is the denom
  name: string
  display: string
  symbol: string
  logo_URIs: {
    png: string
    svg?: string
  }
  coingecko_id?: string
}
