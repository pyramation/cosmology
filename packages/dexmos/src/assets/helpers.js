// need to figure out typescript compatability
// import { Coin } from '../dexmos-types';
import assetList from './assetlist.json';

// /**
//  * 
//  * @param {string} denom 
//  * @returns {Coin}
//  */
export const getCoinFromDenom = (denom) => {
    for (let asset of assetList.assets) {
        if (asset.base === denom) {
            return asset;
        }
    }
}