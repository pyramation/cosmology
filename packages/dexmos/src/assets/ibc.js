// https://github.com/osmosis-labs/osmosis-frontend/blob/master/src/config.ts#L531
export const ibcChannels = [
	{
		counterpartyChainId: 'cosmoshub-4',
		sourceChannelId: 'channel-0',
		destChannelId: 'channel-141',
		coinMinimalDenom: 'uatom',
	},
	{
		counterpartyChainId: 'columbus-5',
		sourceChannelId: 'channel-72',
		destChannelId: 'channel-1',
		coinMinimalDenom: 'uluna',
	},
	{
		counterpartyChainId: 'crypto-org-chain-mainnet-1',
		sourceChannelId: 'channel-5',
		destChannelId: 'channel-10',
		coinMinimalDenom: 'basecro',
	},
	{
		counterpartyChainId: 'columbus-5',
		sourceChannelId: 'channel-72',
		destChannelId: 'channel-1',
		coinMinimalDenom: 'uusd',
	},
	{
		counterpartyChainId: 'secret-4',
		sourceChannelId: 'channel-88',
		destChannelId: 'channel-1',
		coinMinimalDenom: 'uscrt',
	},
	{
		counterpartyChainId: 'juno-1',
		sourceChannelId: 'channel-42',
		destChannelId: 'channel-0',
		coinMinimalDenom: 'ujuno',
	},
	{
		counterpartyChainId: 'stargaze-1',
		sourceChannelId: 'channel-75',
		destChannelId: 'channel-0',
		coinMinimalDenom: 'ustars',
	},
	{
		counterpartyChainId: 'chihuahua-1',
		sourceChannelId: 'channel-113',
		destChannelId: 'channel-7',
		coinMinimalDenom: 'uhuahua',
	},
	{
		counterpartyChainId: 'core-1',
		sourceChannelId: 'channel-4',
		destChannelId: 'channel-6',
		coinMinimalDenom: 'uxprt',
	},
	{
		counterpartyChainId: 'columbus-5',
		sourceChannelId: 'channel-72',
		destChannelId: 'channel-1',
		coinMinimalDenom: 'ukrw',
	},
	{
		counterpartyChainId: 'akashnet-2',
		sourceChannelId: 'channel-1',
		destChannelId: 'channel-9',
		coinMinimalDenom: 'uakt',
	},
	{
		counterpartyChainId: 'regen-1',
		sourceChannelId: 'channel-8',
		destChannelId: 'channel-1',
		coinMinimalDenom: 'uregen',
	},
	{
		counterpartyChainId: 'sentinelhub-2',
		sourceChannelId: 'channel-2',
		destChannelId: 'channel-0',
		coinMinimalDenom: 'udvpn',
	},
	{
		counterpartyChainId: 'irishub-1',
		sourceChannelId: 'channel-6',
		destChannelId: 'channel-3',
		coinMinimalDenom: 'uiris',
	},
	{
		counterpartyChainId: 'iov-mainnet-ibc',
		sourceChannelId: 'channel-15',
		destChannelId: 'channel-2',
		coinMinimalDenom: 'uiov',
	},
	{
		counterpartyChainId: 'emoney-3',
		sourceChannelId: 'channel-37',
		destChannelId: 'channel-0',
		coinMinimalDenom: 'ungm',
	},
	{
		counterpartyChainId: 'emoney-3',
		sourceChannelId: 'channel-37',
		destChannelId: 'channel-0',
		coinMinimalDenom: 'eeur',
	},
	{
		counterpartyChainId: 'microtick-1',
		sourceChannelId: 'channel-39',
		destChannelId: 'channel-16',
		coinMinimalDenom: 'utick',
	},
	{
		counterpartyChainId: 'likecoin-mainnet-2',
		sourceChannelId: 'channel-53',
		destChannelId: 'channel-3',
		coinMinimalDenom: 'nanolike',
	},
	{
		counterpartyChainId: 'impacthub-3',
		sourceChannelId: 'channel-38',
		destChannelId: 'channel-4',
		coinMinimalDenom: 'uixo',
	},
	{
		counterpartyChainId: 'bitcanna-1',
		sourceChannelId: 'channel-51',
		destChannelId: 'channel-1',
		coinMinimalDenom: 'ubcna',
	},
	{
		counterpartyChainId: 'bitsong-2b',
		sourceChannelId: 'channel-73',
		destChannelId: 'channel-0',
		coinMinimalDenom: 'ubtsg',
	},
	{
		counterpartyChainId: 'kichain-2',
		sourceChannelId: 'channel-77',
		destChannelId: 'channel-0',
		coinMinimalDenom: 'uxki',
	},
	{
		counterpartyChainId: 'panacea-3',
		sourceChannelId: 'channel-82',
		destChannelId: 'channel-1',
		coinMinimalDenom: 'umed',
	},
	{
		counterpartyChainId: 'bostrom',
		sourceChannelId: 'channel-95',
		destChannelId: 'channel-2',
		coinMinimalDenom: 'boot',
	},
	{
		counterpartyChainId: 'comdex-1',
		sourceChannelId: 'channel-87',
		destChannelId: 'channel-1',
		coinMinimalDenom: 'ucmdx',
	},
	{
		counterpartyChainId: 'cheqd-mainnet-1',
		sourceChannelId: 'channel-108',
		destChannelId: 'channel-0',
		coinMinimalDenom: 'ncheq',
	},
	{
		counterpartyChainId: 'lum-network-1',
		sourceChannelId: 'channel-115',
		destChannelId: 'channel-3',
		coinMinimalDenom: 'ulum',
	},
	{
		counterpartyChainId: 'vidulum-1',
		sourceChannelId: 'channel-124',
		destChannelId: 'channel-0',
		coinMinimalDenom: 'uvdl',
	},
	{
		counterpartyChainId: 'desmos-mainnet',
		sourceChannelId: 'channel-135',
		destChannelId: 'channel-2',
		coinMinimalDenom: 'udsm',
	},
	{
		counterpartyChainId: 'dig-1',
		sourceChannelId: 'channel-128',
		destChannelId: 'channel-1',
		coinMinimalDenom: 'udig',
	},
];