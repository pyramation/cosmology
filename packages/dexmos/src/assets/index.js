import json from './assetlist.json';
export const assets = json.assets;
// https://raw.githubusercontent.com/osmosis-labs/assetlists/main/osmosis-1/osmosis-1.assetlist.json
