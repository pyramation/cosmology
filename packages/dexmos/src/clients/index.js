export * from './osmosis';
export * from './cosmos';