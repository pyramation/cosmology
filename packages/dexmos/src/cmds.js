
import _create_mnemonic_wallet from './commands/create-mnemonic-wallet';
import _delegate from './commands/delegate';
import _get_balances from './commands/get-balances';
import _get_latest_block from './commands/get-latest-block';
import _get_pool_apr from './commands/get-pool-apr';
import _get_price from './commands/get-price';
import _get_staking_rewards from './commands/get-staking-rewards';
import _get_transaction from './commands/get-transaction';
import _ibc_transfer from './commands/ibc-transfer';
import _ibc_wip from './commands/ibc-wip';
import _join_pool from './commands/join-pool';
import _list_apis from './commands/list-apis';
import _list_pools from './commands/list-pools';
import _list_prices from './commands/list-prices';
import _load_recipe from './commands/load-recipe';
import _lock_tokens from './commands/lock-tokens';
import _send from './commands/send';
import _swap_exact_amount_in from './commands/swap-exact-amount-in';
import _wallet_from_mnemonic from './commands/wallet-from-mnemonic';
const Commands = {};
Commands['create-mnemonic-wallet'] = _create_mnemonic_wallet;
Commands['delegate'] = _delegate;
Commands['get-balances'] = _get_balances;
Commands['get-latest-block'] = _get_latest_block;
Commands['get-pool-apr'] = _get_pool_apr;
Commands['get-price'] = _get_price;
Commands['get-staking-rewards'] = _get_staking_rewards;
Commands['get-transaction'] = _get_transaction;
Commands['ibc-transfer'] = _ibc_transfer;
Commands['ibc-wip'] = _ibc_wip;
Commands['join-pool'] = _join_pool;
Commands['list-apis'] = _list_apis;
Commands['list-pools'] = _list_pools;
Commands['list-prices'] = _list_prices;
Commands['load-recipe'] = _load_recipe;
Commands['lock-tokens'] = _lock_tokens;
Commands['send'] = _send;
Commands['swap-exact-amount-in'] = _swap_exact_amount_in;
Commands['wallet-from-mnemonic'] = _wallet_from_mnemonic;

  export { Commands }; 

  