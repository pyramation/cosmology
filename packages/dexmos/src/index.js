export * from './assets';
export * from './clients';
export * from './utils/osmo';
export * from './driver/driver';
export * from './messages';