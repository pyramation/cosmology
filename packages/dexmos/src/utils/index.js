export * from './chain';
export * from './prompt';
export * from './helpers';
export * from './wallet';
export * from './prompts/osmo';
