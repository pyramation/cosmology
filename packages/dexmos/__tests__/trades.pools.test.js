// @ts-nocheck
import pricesFixture from '../__fixtures__/coingecko/api/v3/simple/price/data.json';
import bankFixture from '../__fixtures__/keplr/bank/balances/osmo1x/data.json';
import poolsFixture from '../__fixtures__/keplr/osmosis/gamm/v1beta1/pools/data.json';
import lockedPoolsFixture from '../__fixtures__/keplr/osmosis/lockup/v1beta1/account_locked_coins/osmo1/data.json';
import { Token } from '../src/model/Token';
import { assets, chains } from '@pyramation/cosmos-registry';
import { assets as osmosisAssets } from '../src/assets';
import cases from 'jest-in-case';
import { displayUnitsToDenomUnits, baseUnitsToDisplayUnits, getOsmosisSymbolIbcName } from '../src/utils';
import Long from 'long';
import {
    getCoinGeckoIdForSymbol,
    getSymbolForCoinGeckoId,
    osmoDenomToSymbol,
    symbolToOsmoDenom,
    OsmosisToken,
    convertCoinToDisplayValues,
    convertCoinsToDisplayValues,
    calculateCoinsTotalBalance,
    convertGeckoPricesToDenomPriceHash,
    getPoolByGammName,
    getUserPools,
    convertPoolToDisplayValues,
    convertPoolsToDisplayValues,
    getFilteredPoolsWithValues,
    symbolsAndDisplayValuesToCoinsArray,
    getTradesRequiredToGetBalances,
    canonicalizeCoinWeights,
    convertWeightsIntoCoins,
    poolAllocationToCoinsNeeded
} from '../src/utils/osmo';

/*
 - [ ] list of coins you want (via weights)
 - [ ] list of pools you want (via weights)

 - [ ] calculate coins needed for pools
 - [ ] calculate routes for swaps
 */

const prices = convertGeckoPricesToDenomPriceHash(pricesFixture);
const pools = getFilteredPoolsWithValues({ prices, pools: poolsFixture.pools });


cases('getPoolByGammName', opts => {
    const prices = convertGeckoPricesToDenomPriceHash(pricesFixture)
    const pools = getFilteredPoolsWithValues({ prices, pools: poolsFixture.pools })
    expect(getPoolByGammName(pools, opts.name).id).toEqual(opts.poolId);
}, [
    {
        name: 'gamm/pool/1',
        poolId: "1"
    },
    {
        name: 'gamm/pool/600',
        poolId: "600"
    },
    {
        name: 'gamm/pool/606',
        poolId: "606"
    }
]);

cases('canonicalizeCoinWeights', opts => {
    expect(canonicalizeCoinWeights({
        weights: opts.weights, pools, prices
    })).toMatchSnapshot();
}, [
    {
        name: 'many weights',
        weights: [
            {
                weight: 5,
                denom: 'gamm/pool/3'
            },
            {
                weight: 5,
                denom: 'gamm/pool/1'
            },
            {
                weight: 5,
                poolId: '600'
            },
            {
                weight: 5,
                denom: 'gamm/pool/606'
            },
            {
                weight: 2,
                symbol: 'LUNA',
                denom: 'ibc/0EF15DF2F02480ADE0BB6E85D9EBB5DAEA2836D3860E9F97F9AADE4F57A31AA0'
            },
            {
                weight: 10,
                symbol: 'UST',
                denom: 'ibc/BE1BB42D4BE3C30D50B68D7C41DB4DFCE9678E8EF8C539F6E6A9345048894FCC'
            }
        ]
    },
    {
        name: '3 weights',
        weights: [
            {
                weight: 5,
                denom: 'gamm/pool/606'
            },
            {
                weight: 2,
                symbol: 'LUNA'
            },
            {
                weight: 10,
                denom: 'ibc/BE1BB42D4BE3C30D50B68D7C41DB4DFCE9678E8EF8C539F6E6A9345048894FCC'
            }
        ]
    },
    {
        name: 'pools only',
        weights: [
            {
                weight: 5,
                denom: 'gamm/pool/3'
            },
            {
                weight: 5,
                denom: 'gamm/pool/1'
            },
            {
                weight: 5,
                poolId: '600'
            }
        ]
    }
]);

it('allocations', async () => {
    const allocation = poolAllocationToCoinsNeeded({
        pools, prices, weight: {
            "allocation": 0.15625,
            "denom": "gamm/pool/600",
            "name": "ATOM/CMDX",
            "poolId": "600",
            "type": "pool",
            "value": 1994.21875,
            "weight": 5,
        }
    });
    expect(allocation).toMatchSnapshot();
});

describe('pool allocations', () => {
    // CONVERT WEIGHTS of POOLS/COINS into COINS

    const balances = symbolsAndDisplayValuesToCoinsArray(
        [
            {
                symbol: 'ATOM',
                amount: 100
            },
            {
                symbol: 'OSMO',
                amount: 1000
            }
        ]
    );
    
    
    const weights = [
        {
            weight: 5,
            denom: 'gamm/pool/3'
        },
        {
            weight: 5,
            denom: 'gamm/pool/1'
        },
        {
            weight: 5,
            poolId: '600'
        },
        {
            weight: 5,
            denom: 'gamm/pool/606'
        },
        {
            weight: 2,
            symbol: 'LUNA',
            denom: 'ibc/0EF15DF2F02480ADE0BB6E85D9EBB5DAEA2836D3860E9F97F9AADE4F57A31AA0'
        },
        {
            weight: 10,
            symbol: 'UST',
            denom: 'ibc/BE1BB42D4BE3C30D50B68D7C41DB4DFCE9678E8EF8C539F6E6A9345048894FCC'
        }
    ];

    it('determines pool allocations', async () => {
        // - [x] determine value of each allocation
        // - [x] given value of pool allocation, determine amount of coins
        const result = convertWeightsIntoCoins({ weights, pools, prices, balances });
        expect(result).toMatchSnapshot();
    });



    it('pool trades', async () => {
        // - [x] determine trades required for pools
        const result = convertWeightsIntoCoins({ weights, pools, prices, balances });
        for (let i=0; i<result.pools.length; i++) {
            const desired = result.pools[i].coins;
            const trades = getTradesRequiredToGetBalances({ prices, balances, desired })
            const a = {
                name: result.pools[i].name,
                trades
            };
            expect(a).toMatchSnapshot();
        }
    });

    it('determines pool allocations', async () => {
        // - [x] determine trades required for coins
        const result = convertWeightsIntoCoins({ weights, pools, prices, balances });
        const trades = getTradesRequiredToGetBalances({ prices, balances, desired: result.coins })
        expect(trades).toMatchSnapshot();
    });
});
