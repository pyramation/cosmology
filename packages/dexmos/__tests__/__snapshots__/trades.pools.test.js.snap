// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse weights 1`] = `
Array [
  Object {
    "allocation": 0.15625,
    "denom": "gamm/pool/3",
    "name": "AKT/OSMO",
    "poolId": "3",
    "type": "pool",
    "weight": 5,
  },
  Object {
    "allocation": 0.15625,
    "denom": "gamm/pool/1",
    "name": "ATOM/OSMO",
    "poolId": "1",
    "type": "pool",
    "weight": 5,
  },
  Object {
    "allocation": 0.15625,
    "denom": "gamm/pool/600",
    "name": "ATOM/CMDX",
    "poolId": "600",
    "type": "pool",
    "weight": 5,
  },
  Object {
    "allocation": 0.15625,
    "denom": "gamm/pool/606",
    "name": "ATOM/HUAHUA",
    "poolId": "606",
    "type": "pool",
    "weight": 5,
  },
  Object {
    "allocation": 0.0625,
    "denom": "ibc/0EF15DF2F02480ADE0BB6E85D9EBB5DAEA2836D3860E9F97F9AADE4F57A31AA0",
    "name": "LUNA",
    "symbol": "LUNA",
    "type": "coin",
    "weight": 2,
  },
  Object {
    "allocation": 0.3125,
    "denom": "ibc/BE1BB42D4BE3C30D50B68D7C41DB4DFCE9678E8EF8C539F6E6A9345048894FCC",
    "name": "UST",
    "symbol": "UST",
    "type": "coin",
    "weight": 10,
  },
]
`;

exports[`pools desired 1`] = `
Array [
  Object {
    "allocation": 0.15625,
    "denom": "gamm/pool/3",
    "name": "AKT/OSMO",
    "poolId": "3",
    "type": "pool",
    "value": 1994.21875,
    "weight": 5,
  },
  Object {
    "allocation": 0.15625,
    "denom": "gamm/pool/1",
    "name": "ATOM/OSMO",
    "poolId": "1",
    "type": "pool",
    "value": 1994.21875,
    "weight": 5,
  },
  Object {
    "allocation": 0.15625,
    "denom": "gamm/pool/600",
    "name": "ATOM/CMDX",
    "poolId": "600",
    "type": "pool",
    "value": 1994.21875,
    "weight": 5,
  },
  Object {
    "allocation": 0.15625,
    "denom": "gamm/pool/606",
    "name": "ATOM/HUAHUA",
    "poolId": "606",
    "type": "pool",
    "value": 1994.21875,
    "weight": 5,
  },
  Object {
    "allocation": 0.0625,
    "denom": "ibc/0EF15DF2F02480ADE0BB6E85D9EBB5DAEA2836D3860E9F97F9AADE4F57A31AA0",
    "name": "LUNA",
    "symbol": "LUNA",
    "type": "coin",
    "value": 797.6875,
    "weight": 2,
  },
  Object {
    "allocation": 0.3125,
    "denom": "ibc/BE1BB42D4BE3C30D50B68D7C41DB4DFCE9678E8EF8C539F6E6A9345048894FCC",
    "name": "UST",
    "symbol": "UST",
    "type": "coin",
    "value": 3988.4375,
    "weight": 10,
  },
]
`;
